[TextField] Add a select mode

ctavan opened the pull request:
Closes #8270.

ctavan commented:
@oliviertassinari this is a very early WIP but I would be thankful about your feedback and whether you think the direction (i.e. including the Select within TextField) is right.

One thing I already stumbled upon is that, unlike with normal text inputs, it doesn't work for me to specify the `onChange` prop directly on the `TextField`, that somehow doesn't get passed through. Any idea why that works for text inputs but not for `Select`?

oliviertassinari commented on docs/src/pages/demos/text-fields/TextFields.js:
```diff
@@ -140,6 +141,29 @@ class TextFields extends React.Component {
+          InputProps={{
+            onChange: this.handleChange('currency'),
+          }}
+          options={[{
```
Oh you even want to abstract the options. Why not accepting children instead?

ctavan commented:
I agree, options should probably not be abstracted away. I have changed it to accept children.

However I still have quite a few design questions:

* Right now I've introduced a boolean flag `select` that causes a `Select` to be rendered instead of an `Input`. Any idea how to make that more future-proof?
* An alternative could be to instead add an `InputComponent` prop directly which would accept `Input` and `Select`? I guess I would have to use `React.cloneElement` then.
* In the select case currently the `InputProps` are spread into the `Select` element. Should this be made more explicit by introducing a new `SelectProps` prop that would be mutex with the `InputProps`? However I see us ending up adding many new props to the TextField API…
* My initial question of on the `onChange` behavior also still remains.

Apart from these questions I already find it quite functional. I'm curious about your feedback and I'll be happy to polish this further according to that feedback.

oliviertassinari commented:
> Right now I've introduced a boolean flag select that causes a Select to be rendered instead of an Input. Any idea how to make that more future-proof?

This sounds like a good idea. I was also thinking of using a `type="select"` but it could be confused with the native meaning. Do we need it to be more future-proof? Do we want the text field to support something more than an Input or a Select? Well, I guess. I think that we can wait for this use case to introduce a breaking change in order to support it. I can't think of a good enum property name for it. Maybe `fieldType`?

> An alternative could be to instead add an InputComponent prop directly which would accept Input and Select? I guess I would have to use React.cloneElement then.

I think that asking people to provide the Select component would be against the TextField "one size fits all" purpose of the component.

> In the select case currently the InputProps are spread into the Select element. Should this be made more explicit by introducing a new SelectProps prop that would be mutex with the InputProps? However I see us ending up adding many new props to the TextField API…

I like the idea of adding a new `SelectProps` property. However, I would change the implementation. The Select component is a wrapper on top of the Input component that injects his own input: the SelectInput component.
So the Select component accepts an `input` property. We could wrap the existing Input of the Textfield when `select={true}` or `fieldType="select"` (depends on what you prefer). Something like this:
```jsx
<Select
  input={
    <Input
      autoComplete={autoComplete}
      autoFocus={autoFocus}
      className={InputClassName}
      defaultValue={defaultValue}
      disabled={disabled}
      multiline={multiline}
      name={name}
      rows={rows}
      rowsMax={rowsMax}
      type={type}
      value={value}
      id={id}
      inputProps={inputProps}
      inputRef={inputRef}
      placeholder={placeholder}
      {...InputProps}
    />
  }
  {...SelectProps}
/>
```

> My initial question of on the onChange behavior also still remains.

Wow, this is an easter egg! It's crazy nobody has been asking to add the `onChange` property before. I think that the TextField should forward it to the right component.

oliviertassinari commented:
Let me know what you think above the previous point. Given that most people will start with this textfield component, over the lower level ones. We need to get the API right 😁 .

ctavan commented:
> This sounds like a good idea. I was also thinking of using a type="select" but it could be confused with the native meaning. Do we need it to be more future-proof? Do we want the text field to support something more than an Input or a Select? Well, I guess. I think that we can wait for this use case to introduce a breaking change in order to support it. I can't think of a good enum property name for it. Maybe fieldType?

We already have the `multiline` flag which renders a `textarea` instead of an `input`, so I guess a `select` flag is in line with that concept. I also have a hard time coming up with a better name for now. The only alternative I can think of would be `enum`, but I don't think that name is a huge improvement.

> I think that asking people to provide the Select component would be against the TextField "one size fits all" purpose of the component.

OK.

> I like the idea of adding a new SelectProps property. However, I would change the implementation. The Select component is a wrapper on top of the Input component that injects his own input: the SelectInput component.
So the Select component accepts an input property. We could wrap the existing Input of the Textfield when select={true} or fieldType="select" (depends on what you prefer). Something like this:

I will give that a try.

> Wow, this is an easter egg! It's crazy nobody has been asking to add the onChange property before. I think that the TextField should forward it to the right component.

What I just don't understand is why everything works as expected for regular `TextFields`, do you have an idea?

ctavan commented:
@oliviertassinari I've changed the interface to accept `SelectProps` and did the `Input`-wrapping as you suggest and I think it feels much better now :).

For now my only remaining question is regarding the `onChange` behavior. Why does that work for normal `Input`-based `TextFields` but not for the `Select`-based text fields.

As far as I understand the code, the `onChange` prop of a `TextField` is passed on to the `FormControl` through the `...other` here:

https://github.com/callemall/material-ui/blob/6107db2c900d67a5259511f9892f0977d1e734af/src/TextField/TextField.js#L175-L182

But then in the `FormControl` I would assume the `onChange` prop is forwarded to the `ComponentProp`:

https://github.com/callemall/material-ui/blob/6107db2c900d67a5259511f9892f0977d1e734af/src/Form/FormControl.js#L194-L225

So I do quite well understand, why the `onChange` event doesn't work for the `Select`-based `TextField`, I just don't understand why/how the `onChange` prop is ever forwarded to the `Input` when specified on a "normal" `TextField`, like in the first example [on the demo page](https://material-ui-1dab0.firebaseapp.com/demos/text-fields/)?!

oliviertassinari commented:
> Why does that work for normal Input-based TextFields but not for the Select-based text fields.

First, the Input can be uncontrolled while the non-native version of the Select component needs to be controlled.
Second, in the case of the Input, we rely on a native input, hence the onChange event bubble up in the tree. I'm gonna assume the FormControl component catch it.

So, I think that we can remove the indirection and apply the onChange property directly on the input component.

oliviertassinari commented:
Hold on, I'm continuing this PR :)

oliviertassinari commented:
We just need some small tweaks to merge it 💃 .

oliviertassinari commented:
@ctavan Great work :)

ctavan commented:
Awesome! Could have continued on Monday but even better to have this merged already. Thanks a lot!
